File: support/src/julia/utils/FileUtils.jl
"""
FileUtils - File system utilities.

Provides common file operations for scripts and modules.

# Exports
- `ensure_directory`: Create directory if it doesn't exist
- `copy_if_newer`: Copy file only if source is newer
- `find_files`: Find files matching a pattern
- `safe_write`: Write with atomic replacement
"""
module FileUtils

export ensure_directory, ensure_parent_directory, copy_if_newer, find_files, safe_write

"""
    ensure_directory(path::AbstractString) -> String

Create directory and all parent directories if they don't exist.
Returns the absolute path.
"""
function ensure_directory(path::AbstractString)::String
    abs_path = abspath(path)
    mkpath(abs_path)
    return abs_path
end

"""
    ensure_parent_directory(path::AbstractString) -> String

Ensure the parent directory of a file exists.
Returns the absolute path to the parent.
"""
function ensure_parent_directory(path::AbstractString)::String
    parent = dirname(abspath(path))
    mkpath(parent)
    return parent
end

"""
    copy_if_newer(source::AbstractString, target::AbstractString; quiet::Bool=false) -> Bool

Copy source to target only if source is newer or target doesn't exist.
Returns true if copy was performed.
"""
function copy_if_newer(source::AbstractString, target::AbstractString; quiet::Bool=false)::Bool
    if !isfile(source)
        quiet || @warn "Source file does not exist: $source"
        return false
    end
    
    if !isfile(target) || mtime(source) > mtime(target)
        ensure_parent_directory(target)
        cp(source, target, force=true)
        quiet || println("ℹ️  Copied: $source → $target")
        return true
    end
    return false
end

"""
    find_files(directory::AbstractString, pattern::Regex; recursive::Bool=true) -> Vector{String}

Find all files in directory matching the pattern. Returns absolute paths.
"""
function find_files(directory::AbstractString, pattern::Regex; recursive::Bool=true)::Vector{String}
    matches = String[]
    abs_dir = abspath(directory)
    
    if recursive
        for (root, dirs, files) in walkdir(abs_dir)
            for f in files
                if occursin(pattern, f)
                    push!(matches, joinpath(root, f))
                end
            end
        end
    else
        for f in readdir(abs_dir)
            if isfile(joinpath(abs_dir, f)) && occursin(pattern, f)
                push!(matches, joinpath(abs_dir, f))
            end
        end
    end
    
    return matches
end

"""
    safe_write(path::AbstractString, content::AbstractString) -> Nothing

Write content to file atomically (write to temp, then rename).
"""
function safe_write(path::AbstractString, content::AbstractString)
    abs_path = abspath(path)
    temp_path = abs_path * ".tmp"
    ensure_parent_directory(abs_path)
    
    open(temp_path, "w") do io
        write(io, content)
    end
    
    mv(temp_path, abs_path, force=true)
    return nothing
end

end # module

File: support/src/python/utils/file_utils.py
"""
file_utils - File system utilities.

Provides common file operations with absolute path handling.
"""

import os
import shutil
import logging
from pathlib import Path
from typing import List, Pattern
import re

logger = logging.getLogger(__name__)


def ensure_directory(path: str) -> str:
    """
    Create directory and all parent directories if they don't exist.
    Returns the absolute path.
    """
    abs_path = os.path.abspath(path)
    os.makedirs(abs_path, exist_ok=True)
    return abs_path


def ensure_parent_directory(path: str) -> str:
    """
    Ensure the parent directory of a file exists.
    Returns the absolute path to the parent.
    """
    abs_path = os.path.abspath(path)
    parent = os.path.dirname(abs_path)
    os.makedirs(parent, exist_ok=True)
    return parent


def copy_if_newer(source: str, target: str, quiet: bool = False) -> bool:
    """
    Copy source to target only if source is newer or target doesn't exist.
    """
    source_path = Path(os.path.abspath(source))
    target_path = Path(os.path.abspath(target))
    
    if not source_path.exists():
        if not quiet:
            logger.warning(f"⚠️  Source file does not exist: {source}")
        return False
    
    if not target_path.exists() or source_path.stat().st_mtime > target_path.stat().st_mtime:
        ensure_parent_directory(str(target_path))
        shutil.copy2(str(source_path), str(target_path))
        if not quiet:
            logger.info(f"ℹ️  Copied: {source} → {target}")
        return True
    
    return False


def find_files(directory: str, pattern: str, recursive: bool = True) -> List[str]:
    """
    Find all files in directory matching the pattern. Returns absolute paths.
    """
    matches = []
    regex = re.compile(pattern)
    abs_dir = os.path.abspath(directory)
    
    if recursive:
        for root, dirs, files in os.walk(abs_dir):
            for f in files:
                if regex.search(f):
                    matches.append(os.path.join(root, f))
    else:
        for f in os.listdir(abs_dir):
            full_path = os.path.join(abs_dir, f)
            if os.path.isfile(full_path) and regex.search(f):
                matches.append(full_path)
    
    return matches


def safe_write(path: str, content: str) -> None:
    """
    Write content to file atomically.
    """
    abs_path = os.path.abspath(path)
    temp_path = abs_path + ".tmp"
    ensure_parent_directory(abs_path)
    
    with open(temp_path, 'w') as f:
        f.write(content)
    
    shutil.move(temp_path, abs_path)
